rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java

```diff
@@ -0,0 +1,897 @@
+import io.narayana.lra.LRAData;
+import io.narayana.lra.client.NarayanaLRAClient;
+import io.narayana.lra.coordinator.api.Coordinator;
+import io.narayana.lra.coordinator.domain.service.LRAService;
+import io.narayana.lra.coordinator.internal.LRARecoveryModule;
+import io.narayana.lra.filter.ServerLRAFilter;
+import io.narayana.lra.logging.LRALogger;
+import io.narayana.lra.provider.ParticipantStatusOctetStreamProvider;
+import org.eclipse.microprofile.lra.annotation.AfterLRA;
+import org.eclipse.microprofile.lra.annotation.Compensate;
+import org.eclipse.microprofile.lra.annotation.Complete;
+import org.eclipse.microprofile.lra.annotation.Forget;
+import org.eclipse.microprofile.lra.annotation.LRAStatus;
+import org.eclipse.microprofile.lra.annotation.ParticipantStatus;
+import org.eclipse.microprofile.lra.annotation.ws.rs.LRA;
+import org.jboss.resteasy.plugins.server.undertow.UndertowJaxrsServer;
+import org.jboss.resteasy.test.TestPortProvider;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+
+import javax.ws.rs.ApplicationPath;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Application;
+import javax.ws.rs.core.Link;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.StringTokenizer;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import static io.narayana.lra.LRAConstants.COORDINATOR_PATH_NAME;
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_CONTEXT_HEADER;
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_PARENT_CONTEXT_HEADER;
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_RECOVERY_HEADER;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class LRATest {
```

Comment:
The answer to  both questions (CDI and making the tests integration tests) is as follows:

I deliberately created the tests under the coordinator because they are unit tests (mvn test -Dmaven.surefire.debug) not integration tests) specifically created for testing coordinator functionality (and it makes the change/debug/test cycle much simpler). If we were to run the tests in other modules then when I change the coordinator I have to switch to other modules do a rebuild of multiple modules and re-run the tests and that is a longer development cycle.

Regarding CDI, the service is logically part of LRARecoveryModule (which is initialised once on startup) so your argument about CDI would equally apply to the LRARecoveryModule but I am sure you would not suggest we initialise recovery modules via CDI. TBH I would prefer to completely remove any dependency on CDI from the core coordinator since it adds unnecessary additional complexity to defining dependencies and having to configure containers - it only got introduced because I was outvoted.

Reply:
But I agree that moving LRACoordinatorRecovery1TestCase and LRACoordinatorRecovery2TestCase over to `lra/test` would make very good sense (although I was thinking about removing them since I have ported them to become unit tests).

Reply:
Oh, I see, my bad. Thanks for explanation. As those are the unit tests then it makes perfect sense. As you ported the second one to be unit test, I would be for removing it from the integration tests.
Having the first one as integration test and move it to `lra/test/basic` would be nice but as you mention that things with byteman are complicated to be ported. I was thinking about it before but it's not easy.
I think this is a topic which would be good to discuss with @jmfinelli as he is working on changes for the `LRACoordinatorRecovery*TestCases`. Maybe he could try to proceed with moving the first test under `lra/test/basic` and adding the jdbc storage working? (Which will be even harder as it has to be configured with Quarkus too. But it would be nice.

Reply:
@mmusgrov I like the idea of having the smallest number of dependencies possible. But here I think we should agree on desing approach. By me it would be good if we strive having lra-client without dependencies like CDI as then it could be easier to port it to other platform (e.g. Spring). For coordinator I'm still thinking about it more as a component which should be using the features provided by MP - e.g. using the MP configuration, you added the fault-tolerance, the CDI is the core aspect of MP and I think it makes sense to use it.

ad `LRARecoveryModule`) for me it's a plugin of the recovery processing in Narayana. The Narayana's plugin system is done without CDI thus there is no reason to have it as CDI bean (still CDI could be good to be used as a plugin system).

When the CDI would not be used then I think the `LRAService` is not good to be intialized from a recovery module. It's a core class for the transaction/lra processing. The recovery module should not be the place where the `LRAService` is created and stored.

Maybe coordinator could be written without any MP related dependencies and the MP could be moved as addition to a separate module. The it would be good to separate the dependencies completely.

What about to have some brainstorm meeting for clearing the ideas and standpoints around this?

Reply:
These are two good reviews with some valuable insights. On this particular comment my response is that LRAService exposes access to internal classes that implement the coordinator (it is glue between the Coordinator JAX-RS resource and the Arjuna BasicAction/LRA with state maintained by the recovery module and this function limits its' shareability).

Putting it in the bean manager is unnecessary for sharing it with the coordinator and, worse, exposes our internals to any subsystem/extension/etc that wants to look it up via the bean manager (which would fossilise the internal design). As such brainstorming will provide little benefit since we are not looking for solutions, the design is as it is.

Reply:
If it helps, the bottom line is that the life-cycle of LRAService and LRARecoveryModule are identical and the recovery module is the definitive source of information about LRAs and LRAService is just a via.

Reply:
I think that the design can be adjusted if we find the project may benefit from such change. But ok, it was just an idea about brain storming.

I understand the point about the `LRAService` and that could be considered as an internal class. Still CDI may provide a better way to be loosely coupled to other internal components and easier to manage it. I would choose such design solution but as you strongly oppose it I'm fine to go without CDI.

With this I would like to ask your for not `LRAService` being accessible through recovery module. Could we thing having another storage for it? I think the `Coordinator` would be better.

Reply:
I have one more question about testing as I run the tests on my local machine. I found what I haven't understood. We take the term unit test differently. For me the unit test does not involve a runtime. What was changed is that instead of WildFly runtime (and in future the Quarkus) you changed the runtime to Undertow JAX-RS extension. I see the benefit of speed, but is it really that we want to test on Undertow instead on WildFly/Quarkus?

Reply:
I think that the design is adequate as it is. The class is part of the implementation, to restate what I said earlier "the life-cycle of LRAService and LRARecoveryModule are identical and the recovery module is the definitive source of information about LRAs and LRAService is just a via". Your request is weird, LRAService is just exposing the LRAs managed by the implementation.

Reply:
Ok. Let's go with `LRARecoveryModule` then.

Reply:
I still wonder about the tests - do you think running the test with `JAX-RS Undertow`, which is an unsupported container - as a good move?

Reply:
The junit tests are testing the transaction life-cycle via the JAX-RS/HTTP API, that does not make it an integration test, we use undertow to handle http calls so that still does not necessarily mean it's an integration test.

Undertow is the HTTP server used in a number of supported containers such as EAP and quarkus which makes it a "good move" because it means that Undertow is a well tested and actively developed component and Red Hat have active committers on it which will help to guarantee timely bug fixes and technical help should we need it.

We have plenty of arquillian and TCK tests to give us confidence on running on other platforms.

Reply:
I'm not in the agreement with the statement here. Ok, let's not discuss this further. Better some test than nothing.

Reply:
As far as I understand (and correct me if I am wrong), unit tests are those tests that do not depend on external dependencies or modules that can affect the tests. This means that it is possible to "unit test" relatively small pieces of code/functionalities in isolation, supplying mock objects if needed (instead of real objects). Supplying mock objects allows the developer/tester to test the code/functionality avoiding bugs coming from external dependencies/modules.

Whereas, Integration tests are those tests that use external dependencies/modules to create an environment where the testing logic can be proved.

Here is where I think things get complicated: from what I have read, the demarcation line between unit tests and integration tests is often hard to define. Integration tests are a wide categories and it is up to the developer/tester to define whether tests fall in this category or whether they are unit tests, as the line between the two is often blurred. For example, using a in-memory database can be considered to be either an integration test or a unit test.

In this particular situation, it is true that Undertow is used only as a Web Server (to handle HTTP calls) and makes things much easier and faster to test. I suppose, then, that it is up to @mmusgrov to decide in which category this test falls.